duqtools/operations.py

```diff
@@ -0,0 +1,145 @@
+
+import logging
+from collections import deque
+from inspect import signature
+from typing import Callable
+
+import click
+from pydantic import Field
+
+from .schema import BaseModel
+
+logger = logging.getLogger(__name__)
+
+
+class Operation(BaseModel):
+    """Operation, simple class which has a callable action."""
+
+    description: str = Field(
+        description='description of the operation to be done')
+    action: Callable = Field(
+        description='a function which can be executed when we '
+        'decide to apply this operation')
+    args: tuple = Field((),
+                        description='positional arguments that have to be '
+                        'passed to the action')
+    kwargs: dict = Field({},
+                         description='keyword arguments that will be '
+                         'passed to the action')
+
+    def __call__(self) -> Operation:
+        logger.info(f'-: {self.description}')
+        self.action(*self.args, **self.kwargs)
+        return self
+
+
+class Operations(deque):
+    """Operations Queue which keeps track of all the operations that need to be
+    done."""
+
+    _instance = None
+    yes = False  # Apply operations without prompt
+    dry_run = False  # Never apply any operations (do not even ask)
+
+    def __new__(cls, *args, **kwargs):
+        # Make it a singleton
+        if not Operations._instance:
+            Operations._instance = super().__new__(cls)
+        return Operations._instance
+
+    def add(self, *args, **kwargs) -> None:
+        """convenience Operation wrapper around put."""
+
+        self.append(Operation(*args, **kwargs))
+
+    def put(self, item: Operation) -> None:
+        """synonym for append."""
+        self.append(item)
+
+    def append(self, item: Operation) -> None:  # type: ignore
+        """Restrict our diet to Operation objects only."""
+
+        super().append(item)
+
+    def apply(self) -> Operation:
+        """Apply the next operation in the queue and remove it."""
+
+        return self.popleft()()
+
+    def apply_all(self) -> None:
+        """Apply all queued operations and empty the queue."""
+        while len(self) != 0:
+            self.apply()
+
+    def confirm_apply_all(self) -> bool:
+        """First asks the user if he wants to apply everything.
+
+        Returns
+        -------
+        bool: did we apply everything or not
+        """
+
+        # To print the descriptions we need to get them
+        logger.info('')
+        logger.info('Operations in the Queue:')
+        logger.info('========================')
+        for op in self:
+            logging.info(op.description)
+
+        if self.dry_run:
+            logger.info('Dry run enabled, not applying op_queue')
+            return False
+
+        if not self.yes:
+            ans = click.confirm('Do you want to apply all these operations?',
+                                default=False)
+        else:
+            ans = self.yes
```

Comment:
Can't you do something clever here like this? 😅

```suggestion
        ans = self.yes or click.confirm('Do you want to apply all these operations?',
                                default=False)
```

Reply:
even better!